crates/pallet-domains/src/block_tree.rs

```diff
@@ -0,0 +1,716 @@
+use frame_support::{ensure, PalletError};
+use scale_info::TypeInfo;
+use sp_core::Get;
+use sp_domains::v2::ExecutionReceipt;
+use sp_domains::{DomainId, OperatorId};
+use sp_runtime::traits::{CheckedSub, One, Saturating, Zero};
+use sp_std::cmp::Ordering;
+use sp_std::vec::Vec;
+
+/// Block tree specific errors
+#[derive(TypeInfo, Encode, Decode, PalletError, Debug, PartialEq)]
+pub enum Error {
+    InvalidExtrinsicsRoots,
+    UnknownParentBlockReceipt,
+    BuiltOnUnknownConsensusBlock,
+    InFutureReceipt,
+    PrunedReceipt,
+    BadGenesisReceipt,
+    UnexpectedReceiptType,
+    MaxHeadDomainNumber,
+}
+
+#[derive(TypeInfo, Debug, Encode, Decode, Clone, PartialEq, Eq)]
+pub struct DomainBlock<Number, Hash, DomainNumber, DomainHash, Balance> {
+    /// The full ER for this block.
+    pub execution_receipt: ExecutionReceipt<Number, Hash, DomainNumber, DomainHash, Balance>,
+    /// A set of all operators who have committed to this ER within a bundle. Used to determine who to
+    /// slash if a fraudulent branch of the `block_tree` is pruned.
+    ///
+    /// NOTE: there may be duplicated operator id as an operator can submit multiple bundles with the
+    /// same head receipt to a consensus block.
+    pub operator_ids: Vec<OperatorId>,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub(crate) enum AcceptedReceiptType {
+    // New head receipt that extend the longest branch
+    NewHead,
+    // Receipt that creates a new branch of the block tree
+    NewBranch,
+    // Receipt that comfirms the current head receipt
+    CurrentHead,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub(crate) enum RejectedReceiptType {
+    // Receipt that is newer than the head receipt but does not extend the head receipt
+    InFuture,
+    // Receipt that already been pruned
+    Pruned,
+}
+
+impl From<RejectedReceiptType> for Error {
+    fn from(rejected_receipt: RejectedReceiptType) -> Error {
+        match rejected_receipt {
+            RejectedReceiptType::InFuture => Error::InFutureReceipt,
+            RejectedReceiptType::Pruned => Error::PrunedReceipt,
+        }
+    }
+}
+
+/// The type of receipt regarding to its freshness
+#[derive(Debug, PartialEq, Eq)]
+pub(crate) enum ReceiptType {
+    Accepted(AcceptedReceiptType),
+    Rejected(RejectedReceiptType),
+    // Receipt that comfirm a non-head receipt
+    Stale,
+}
+
+/// Get the receipt type of the given receipt based on the current block tree state
+pub(crate) fn execution_receipt_type<T: Config>(
+    domain_id: DomainId,
+    execution_receipt: &ExecutionReceiptOf<T>,
+) -> ReceiptType {
+    let receipt_number = execution_receipt.domain_block_number;
+    let head_receipt_number = HeadReceiptNumber::<T>::get(domain_id);
+
+    match receipt_number.cmp(&head_receipt_number.saturating_add(One::one())) {
+        Ordering::Greater => ReceiptType::Rejected(RejectedReceiptType::InFuture),
+        Ordering::Equal => ReceiptType::Accepted(AcceptedReceiptType::NewHead),
+        Ordering::Less => {
```

Comment:
not a blocker but would be great to improve this arm. One too many `if..else if` IMO

Reply:
`if .. else if` can be replaced by a bunch of `if .. { return Error::xx }` but logically it is the same.

Reply:
🤷🏼‍♂️